File: solutions/scala/nth-prime/src/main/scala/NthPrime.scala
object NthPrime:
  def prime(n: Int): Option[Int] =
    if n < 1 then None
    else Some(calculatePrime(n))

  private def calculatePrime(nth: Int): Int =
    val primes = scala.collection.mutable.ArrayBuffer(2)

    def isPrime(candidate: Int): Boolean =
      primes
        .takeWhile(_ <= Math.sqrt(candidate))
        .forall(candidate % _ != 0)

    LazyList.iterate(3)(_ + 2)
      .takeWhile(_ => primes.size < nth)
      .filter(isPrime)
      .foreach(primes += _)

    primes(nth - 1)

File: solutions/fsharp/nth-prime/NthPrime.fs
module NthPrime

let calculatePrime nth =
    let primes = ResizeArray<int>([ 2 ])
    let isPrime number = primes |> Seq.forall (fun p -> number % p <> 0)

    seq { 3 .. 2 .. 9999999 }
    |> Seq.takeWhile (fun _ -> primes.Count < nth)
    |> Seq.filter isPrime
    |> Seq.iter primes.Add

    primes.Item(nth - 1)

let prime nth: int option =
    if nth < 1 then None else Some <| calculatePrime nth
